openhands/server/routes/git.py

```diff
@@ -229,3 +239,268 @@ async def get_repository_branches(
+    """Get the creation time of a file from Git history.
+
+    Args:
+        repo_dir: The root directory of the Git repository
+        file_path: The path to the file relative to the repository root
+
+    Returns:
+        datetime: The timestamp when the file was first added to the repository
+    """
+    try:
+        # Get the relative path from the repository root
+        relative_path = file_path.relative_to(repo_dir)
+
+        # Use git log to get the first commit that added this file
+        # --follow: follow renames and moves
+        # --reverse: show commits in reverse chronological order (oldest first)
+        # --format=%ct: show commit timestamp in Unix format
+        # -1: limit to 1 result (the first commit)
+        cmd = [
+            'git',
+            'log',
+            '--follow',
+            '--reverse',
+            '--format=%ct',
+            '-1',
+            str(relative_path),
+        ]
+
+        result = subprocess.run(
+            cmd, cwd=repo_dir, capture_output=True, text=True, timeout=10
+        )
+
+        if result.returncode == 0 and result.stdout.strip():
+            # Parse Unix timestamp and convert to datetime
+            timestamp = int(result.stdout.strip())
+            return datetime.fromtimestamp(timestamp)
+        else:
+            logger.warning(
+                f'Failed to get creation time for {relative_path}: {result.stderr}'
+            )
+            # Fallback to current time if git log fails
+            return datetime.now()
+
+    except Exception as e:
+        logger.warning(f'Error getting creation time for {file_path}: {str(e)}')
+        # Fallback to current time if there's any error
+        return datetime.now()
+
+
+@app.get(
+    '/repository/{repository_name:path}/microagents',
+    response_model=list[MicroagentResponse],
+)
+async def get_repository_microagents(
+    repository_name: str,
+    provider_tokens: PROVIDER_TOKEN_TYPE | None = Depends(get_provider_tokens),
+    access_token: SecretStr | None = Depends(get_access_token),
+    user_id: str | None = Depends(get_user_id),
+) -> list[MicroagentResponse] | JSONResponse:
+    """Scan the .openhands/microagents directory of a repository and return the list of microagents.
+
+    Args:
+        repository_name: Repository name in the format 'owner/repo' or 'domain/owner/repo'
+        provider_tokens: Provider tokens for authentication
+        access_token: Access token for external authentication
+        user_id: User ID for authentication
+
+    Returns:
+        List of microagents found in the repository's .openhands/microagents directory
+    """
+    if not provider_tokens:
+        logger.info(
+            f'Returning 401 Unauthorized - Git provider token required for user_id: {user_id}'
+        )
+        return JSONResponse(
+            content='Git provider token required. (such as GitHub).',
+            status_code=status.HTTP_401_UNAUTHORIZED,
+        )
```

Comment:
```suggestion
```

This is the confusing part of our codebase - we don't use provider tokens in cloud openhands (this the token the user passes to us)

We have two options, we can check `if (not provider_tokens and server_config.app_mode == AppMode.OSS) or (not user_id and server_config.app_mode == AppMode.SAAS)`

Alternatively, we can skip this logic all together and the ProviderHandler will throw an error which the try catch will handle